Replace python-nose with unittest

pkratoch opened the pull request:
To check that some tests were run, the script that runs the tests now
checks output instead of the xunit file, because the unittest module
doesn't have the option to generate this file.

Also, the unittest is run with -s (--start-directory) option as well as
-t (--top-level-directory) option, because the "module" module in
python/hawkey/tests/ has cyclic import and unittest errors out when
trying to import it.

= changelog =
resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1916786

pkratoch committed:
Replace python-nose with unittest

The unittest is run with -s (--start-directory) option as well as
-t (--top-level-directory) option, because the "module" module in
python/hawkey/tests/ has cyclic import and unittest errors out when
trying to import it.

= changelog =
resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1916786

pkratoch commented:
I had no idea how to resolve the cyclic import in the ``python/hawkey/tests/module/``, so I used ``-t`` option for unittest as a workaround. If you have any tip how to resolve it, let me know.

dmach commented:
> I had no idea how to resolve the cyclic import in the `python/hawkey/tests/module/`, so I used `-t` option for unittest as a workaround. If you have any tip how to resolve it, let me know.

Maybe try the following approach: https://github.com/rpm-software-management/libdnf/blob/dnf-5-devel/test/python3/libdnf/CMakeLists.txt (drop  python/hawkey/tests/tests/run_unittests.in)
There's no need to have a test runner written in python when you can run the tests directly from cmake.

pkratoch commented:
I dropped the run_unittests.in and run the tests directly from cmake as you suggested.

As a consequence, the check that some tests were run was also removed. The check was introduced by commit https://github.com/rpm-software-management/libdnf/commit/a6d7f6634118acfe84763e8c873b72512498e581 and the bug it references (https://bugzilla.redhat.com/show_bug.cgi?id=1226753) was probably never resolved, but the ``python_rhev`` looks kind of dead, so I think it's safe to remove the workaround.

lukash commented on README.md:
```diff
@@ -20,7 +20,7 @@ Building for Fedora
 To install build requirements, run following command:
 
-    dnf install check-devel cmake cppunit-devel gcc gcc-c++ glib2-devel gpgme-devel gtk-doc json-c-devel libmodulemd-devel librepo-devel libsolv-devel libsolv-tools make python2-devel python3-devel python2-nose python3-nose python2-sphinx python3-sphinx python2-breathe python3-breathe rpm-devel sqlite-devel swig libsmartcols-devel
+    dnf install check-devel cmake cppunit-devel gcc gcc-c++ glib2-devel gpgme-devel gtk-doc json-c-devel libmodulemd-devel librepo-devel libsolv-devel libsolv-tools make python2-devel python3-devel python2-sphinx python3-sphinx python2-breathe python3-breathe rpm-devel sqlite-devel swig libsmartcols-devel
```
I think we should just put a `dnf builddep libdnf.spec` here (potentially prepended by `dnf install dnf-plugins-core` for completeness) instead of a duplicate dependency list that's always bound to be out of date.

lukash commented on python/hawkey/tests/tests/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
+add_test(NAME test_python3_libdnf COMMAND ${PYTHON_EXECUTABLE} -m unittest discover -t "${CMAKE_CURRENT_SOURCE_DIR}/.." -s "${CMAKE_CURRENT_SOURCE_DIR}")
+set_property(TEST test_python3_libdnf PROPERTY ENVIRONMENT
+    "PYTHONPATH=${CMAKE_BINARY_DIR}/bindings/python"
+)
```
This is much better than that script :slightly_smiling_face:

lukash reviewed (commented):
bors try

lukash commented:
bors try

lukash commented:
@pkratoch it seems the unit tests are failing in Copr build: https://download.copr.fedorainfracloud.org/results/rpmsoftwaremanagement/rpm-gitoverlay-1615283368.172544/fedora-33-x86_64/02061299-libdnf/builder-live.log.gz

Can you check it please?

(Also, when you push this, there should be CI in GHA :wink: assuming it works)

pkratoch commented:
It seems I was too hasty with changing the CMakeLists according to the DNF 5 one. Hopefully, I fixed it, let's wait for the CI.

pkratoch commented:
Hm, I am not sure why the CI failed. The process completed with exit code 1, but there are no failed features.

lukash commented:
I have found this in the log:
```
2021-03-09T15:38:47.5743139Z Error: Command "podman run --rm --volume /__w/libdnf/libdnf/test-results:/junit:z fba441ea-a5c6-4bce-a4c5-86280da45e93 behave --dry-run features/plugins-core/debug-restore.feature" failed with 127
2021-03-09T15:38:47.5744492Z stderr:
2021-03-09T15:38:47.5745127Z Error: executable file `behave` not found in $PATH: No such file or directory: OCI not found
```
Pretty hard to find. The `parallel` tool runs the feature files or steps one by one and if one of them fails, the whole thing fails. The failure is just somewhere in the middle...

The failure is weird, I've re-run the CI, let's see what it does.
